### PR TITLE
HBASE-27397 Spark-hbase support for 'startWith' predicate

### PR DESCRIPTION
**Pull request by richardantal**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 27s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 51s |  master passed  |
| +1 :green_heart: |  compile  |   0m 44s |  master passed  |
| +1 :green_heart: |  scaladoc  |   0m 44s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 44s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 36s |  the patch passed  |
| +1 :green_heart: |  scalac  |   0m 36s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  scaladoc  |   0m 45s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   7m 31s |  hbase-spark in the patch passed.  |
|  |   |  15m  0s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-105/1/artifact/yetus-precommit-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase-connectors/pull/105 |
| Optional Tests | dupname scalac scaladoc unit compile |
| uname | Linux 610db1b435a5 5.4.0-1088-aws #96~18.04.1-Ubuntu SMP Mon Oct 17 02:57:48 UTC 2022 x86_64 GNU/Linux |
| Build tool | hb_maven |
| Personality | dev-support/jenkins/hbase-personality.sh |
| git revision | master / 496cd58 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-105/1/testReport/ |
| Max. process+thread count | 942 (vs. ulimit of 12500) |
| modules | C: spark/hbase-spark U: spark/hbase-spark |
| Console output | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-105/1/console |
| versions | git=2.20.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by richardantal** [comment]
@petersomogyi @meszibalu @saintstack Could you please review this change?

File: spark/hbase-spark/src/main/scala/org/apache/hadoop/hbase/spark/datasources/Utils.scala [hunk 1]
```diff
@@ -89,4 +89,47 @@ object Utils {
...
+    while (a >= 0) {
+      if (array(a) != (-1).toByte) {
+        index = a
+        a = -1 // break from the loop because we found a non -1 element
+      }
+      a = a - 1
+    }
+
+    val shouldIncreaseLength = index < 0
+    var newLenght = array.length - 1
```
Comment (meszibalu) [inline spark/hbase-spark/src/main/scala/org/apache/hadoop/hbase/spark/datasources/Utils.scala:110]: `newLength`

File: spark/hbase-spark/src/main/scala/org/apache/hadoop/hbase/spark/datasources/Utils.scala [hunk 2]
```diff
@@ -89,4 +89,47 @@ object Utils {
...
+
+    if (shouldIncreaseLength) {
```
Comment (meszibalu) [inline spark/hbase-spark/src/main/scala/org/apache/hadoop/hbase/spark/datasources/Utils.scala:112]: Why should we increase the length? If someone passes `0xFF` for `startsWith` then it is true that `0x01 0x00` if the next value, however the scan range in `startsWith` going to be `0xFF <= row < 0x10 0x00` which is wrong.

**Review by meszibalu — CHANGES_REQUESTED** [review]

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m  1s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 35s |  master passed  |
| +1 :green_heart: |  compile  |   0m 39s |  master passed  |
| +1 :green_heart: |  scaladoc  |   0m 49s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 47s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 38s |  the patch passed  |
| +1 :green_heart: |  scalac  |   0m 38s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  scaladoc  |   0m 48s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   8m  1s |  hbase-spark in the patch passed.  |
|  |   |  16m  3s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-105/2/artifact/yetus-precommit-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase-connectors/pull/105 |
| Optional Tests | dupname scalac scaladoc unit compile |
| uname | Linux e0be62915d72 5.4.0-124-generic #140-Ubuntu SMP Thu Aug 4 02:23:37 UTC 2022 x86_64 GNU/Linux |
| Build tool | hb_maven |
| Personality | dev-support/jenkins/hbase-personality.sh |
| git revision | master / 496cd58 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-105/2/testReport/ |
| Max. process+thread count | 943 (vs. ulimit of 12500) |
| modules | C: spark/hbase-spark U: spark/hbase-spark |
| Console output | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-105/2/console |
| versions | git=2.20.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by meszibalu** [comment]
I also added some debug information here: [HBaseTableScanRDD.scala:90](https://github.com/apache/hbase-connectors/blob/master/spark/hbase-spark/src/main/scala/org/apache/hadoop/hbase/spark/datasources/HBaseTableScanRDD.scala#L90)

```scala
if (log.isDebugEnabled) {
    logDebug(s"Partitions: ${ps.size}");
    ps.foreach(x => logDebug(x.toString))
}
```

Then I added some test data with `hbase shell`:

```
create 'spark', 'f', SPLITS => ['m']

put 'spark', 'a1', 'f:value', 'a1'
put 'spark', 'a2', 'f:value', 'a2'
put 'spark', 'b1', 'f:value', 'b1'
put 'spark', 'b2', 'f:value', 'b2'
put 'spark', 'c1', 'f:value', 'c1'
put 'spark', 'c2', 'f:value', 'c2'
put 'spark', 'k1', 'f:value', 'k1'
put 'spark', 'k2', 'f:value', 'k2'
put 'spark', 'l1', 'f:value', 'l1'
put 'spark', 'l2', 'f:value', 'l2'
put 'spark', 'm1', 'f:value', 'm1'
put 'spark', 'm2', 'f:value', 'm2'
put 'spark', 'n1', 'f:value', 'n1'
put 'spark', 'n2', 'f:value', 'n2'
put 'spark', 'o1', 'f:value', 'o1'
put 'spark', 'o2', 'f:value', 'o2'
put 'spark', 'p1', 'f:value', 'p1'
put 'spark', 'p2', 'f:value', 'p2'
```

And run the following spark code:

```scala
org.apache.log4j.LogManager.getLogger("org.apache.hadoop.hbase").setLevel(org.apache.log4j.Level.DEBUG)
import org.apache.spark.sql.functions.col

val sql = spark.sqlContext

val df = sql.read.format("org.apache.hadoop.hbase.spark").option("hbase.columns.mapping", "key STRING :key, value STRING f:value").option("hbase.table", "spark").option("hbase.spark.use.hbasecontext", false).load()
// this is good, uses 2 regions, dumps everything
df.filter(col("key").startsWith("")).show()
// this uses 1 region, but it does not return any row
df.filter(col("key").startsWith("a")).show()
// this uses 1 region and it is ok
df.filter(col("key").equalTo("a1")).show()
```

So `startsWith("a")` does not work for me.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 58s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 44s |  master passed  |
| +1 :green_heart: |  compile  |   0m 39s |  master passed  |
| +1 :green_heart: |  scaladoc  |   0m 47s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 47s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 38s |  the patch passed  |
| +1 :green_heart: |  scalac  |   0m 38s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  scaladoc  |   0m 48s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   8m  3s |  hbase-spark in the patch passed.  |
|  |   |  16m  7s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-105/3/artifact/yetus-precommit-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase-connectors/pull/105 |
| Optional Tests | dupname scalac scaladoc unit compile |
| uname | Linux 2e6b29266b35 5.4.0-124-generic #140-Ubuntu SMP Thu Aug 4 02:23:37 UTC 2022 x86_64 GNU/Linux |
| Build tool | hb_maven |
| Personality | dev-support/jenkins/hbase-personality.sh |
| git revision | master / 496cd58 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-105/3/testReport/ |
| Max. process+thread count | 942 (vs. ulimit of 12500) |
| modules | C: spark/hbase-spark U: spark/hbase-spark |
| Console output | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-105/3/console |
| versions | git=2.20.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**